File: EXAMPLES_Zig/jetgpio_round_trip.zig
// Usage example of the JETGPIO library
// Compile with: zig build-exe jetgpio_round_trip.zig -ljetgpio -lc
// Execute with: sudo ./jetgpio_round_trip

const std = @import("std");
const jetgpio = @cImport({
    @cInclude("jetgpio.h");
});

pub fn main() !void {
    const init = jetgpio.gpioInitialise();

    if (init < 0) {
        std.debug.print("Error initiating jetgpio\n", .{});
    }
    // Setting up pin 38 as output
    const set1 = jetgpio.gpioSetMode(38, jetgpio.JET_OUTPUT);
    if (set1 < 0) {
        std.debug.print("Error setting pin 38\n", .{});
    }

    // Setting up pin 40 as input
    const set2 = jetgpio.gpioSetMode(40, jetgpio.JET_INPUT);
    if (set2 < 0) {
        std.debug.print("Error setting pin 40\n", .{});
    }

    // Measuring time now
    const start = try std.time.Instant.now();

    // Writing 1 to pin 38
    _ = jetgpio.gpioWrite(38, 1);

    while (jetgpio.gpioRead(40) == 0) {}

    // Measuring time after the loop
    const end = try std.time.Instant.now();

    // Calculating time difference for round trip e.g. time invested in activating the output + the time it takes for the input to detect the level change
    const elapsed = end.since(start);
    std.debug.print("Elapsed time = {} nanoseconds\n", .{elapsed});

    // Writing 0 to pin 38
    _ = jetgpio.gpioWrite(38, 1);

    jetgpio.gpioTerminate();
}

File: EXAMPLES_C/jetgpio_round_trip.c
/* Usage example of the JETGPIO library
 * Compile with: gcc -Wall -o jetgpio_round jetgpio_round_trip.c -ljetgpio -lrt
 * Execute with: sudo ./jetgpio_round
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <jetgpio.h>

#define BILLION 1000000000L

int main(int argc, char *argv[])
{
  int Init;
  uint64_t diff;
  struct timespec start, end;

  Init = gpioInitialise();
  if (Init < 0)
    {
      /* jetgpio initialisation failed */
      printf("Jetgpio initialisation failed. Error code:  %d\n", Init);
      exit(Init);
    }
  else
    {
      /* jetgpio initialised okay*/
      printf("Jetgpio initialisation OK. Return code:  %d\n", Init);
    }	

  // Setting up pin 38 as OUTPUT and 40 as INPUT

  gpioSetMode(38, JET_OUTPUT);
  gpioSetMode(40, JET_INPUT);

  // Measuring monotonic time now
  clock_gettime(CLOCK_MONOTONIC, &start);	

  // Writing 1 to pin 38
  gpioWrite(38, 1);

  while (!gpioRead(40)) {}

  // Measuring monotonic time after the loop
  clock_gettime(CLOCK_MONOTONIC, &end);	

  // Calculating time difference for round trip e.g. time invested in activating the output + the time it takes for the input to detect the level
  diff = BILLION * (end.tv_sec - start.tv_sec) + end.tv_nsec - start.tv_nsec;
  printf("elapsed time = %llu nanoseconds\n", (long long unsigned int) diff);

  gpioWrite(38, 0);

  // Terminating library 
  gpioTerminate();

  exit(0);
	
}

